In [1]:
%%capture
pip install pandas

In [4]:
import pandas as pd
import seaborn as sns
import math
import numpy as np

In [6]:
df = pd.read_csv("https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv")

In [7]:
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

## Metode kNN

<h2>Tahap Persiapan</h2>

1.   Mengimport library yang dibutuhkan seperti pandas untuk membaca dataset dan scikit-learn untuk kNN.
2.   Membaca dataset menggunakan pandas dan memisahkan variabel independen dan dependen.
3.   Membagi dataset menjadi training set dan test set.

In [10]:
# Import library yang dibutuhkan
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

# Membaca dataset
data = pd.read_csv('https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv')

# Pemisahan variabel independen dan dependen
X = data.iloc[:, :-1] # Variabel independen
y = data.iloc[:, -1] # Variabel dependen

# Membagi dataset menjadi training set dan test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

<h2>Tahap Pelatihan Model</h2>

4.   Melatih model kNN menggunakan training set dan menentukan nilai k terbaik dengan melakukan validasi silang (cross-validation) menggunakan GridSearchCV.

In [11]:
# Melatih model kNN
knn = KNeighborsClassifier()

# Menentukan nilai k terbaik menggunakan GridSearchCV
from sklearn.model_selection import GridSearchCV
k_range = list(range(1, 31))
param_grid = dict(n_neighbors=k_range)
grid = GridSearchCV(knn, param_grid, cv=10, scoring='accuracy')
grid.fit(X_train, y_train)

GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12,
                                         13, 14, 15, 16, 17, 18, 19, 20, 21, 22,
                                         23, 24, 25, 26, 27, 28, 29, 30]},
             scoring='accuracy')

<h2>Tahap Evaluasi Model</h2>

5.   Menggunakan model kNN yang sudah dilatih untuk melakukan prediksi pada test set dan mengevaluasi performa model menggunakan metrik evaluasi seperti akurasi, presisi, recall, dan F1-score.

In [12]:
# Melakukan prediksi pada test set
y_pred = grid.predict(X_test)

# Evaluasi performa model
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('F1 Score:', f1_score(y_test, y_pred))

Accuracy: 0.6593406593406593
Precision: 0.6792452830188679
Recall: 0.72
F1 Score: 0.6990291262135921


<h2>Hasil Output</h2>

6.   Setelah melakukan tahap persiapan, pelatihan model, dan evaluasi model, maka output yang dihasilkan adalah metrik evaluasi performa model seperti akurasi, presisi, recall, dan F1-score.

Output tersebut menunjukkan bahwa model kNN yang dilatih dengan dataset heart-disease.csv berhasil mencapai akurasi sebesar 65.93%, presisi sebesar 67.92%, recall sebesar 72%, dan F1-score sebesar 69.90%. Namun, nilai-nilai tersebut dapat berbeda-beda tergantung pada pemilihan nilai k yang optimal dan dataset yang digunakan.

## Metode Naive Bayes

<h2>Tahap Persiapan</h2>

1.   Mengimport library yang dibutuhkan seperti pandas untuk membaca dataset dan scikit-learn untuk kNN.
2.   Membaca dataset menggunakan pandas dan memisahkan variabel independen dan dependen.
3.   Membagi dataset menjadi training set dan test set.

In [17]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report

# Load dataset
data = pd.read_csv('https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv')

# Separate features and target variable
X = data.iloc[:, :-1] # Features
y = data.iloc[:, -1] # Target variable

<h2>Tahap Pelatihan Model</h2>

4.   Melatih model kNN menggunakan training set dan menentukan nilai k terbaik dengan melakukan validasi silang (cross-validation) menggunakan GridSearchCV.

In [20]:
# Normalize features if needed

# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train Naive Bayes classifier
clf = GaussianNB()
clf.fit(X_train, y_train)

# Test the classifier
y_pred = clf.predict(X_test)

<h2>Tahap Evaluasi Model</h2>

5.   Menggunakan model kNN yang sudah dilatih untuk melakukan prediksi pada test set dan mengevaluasi performa model menggunakan metrik evaluasi seperti akurasi, presisi, recall, dan F1-score.

In [21]:
# Evaluate the classifier
acc = accuracy_score(y_test, y_pred)
prec = precision_score(y_test, y_pred)
rec = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Accuracy:", acc)
print("Precision:", prec)
print("Recall:", rec)
print("F1-score:", f1)

Accuracy: 0.8688524590163934
Precision: 0.9
Recall: 0.84375
F1-score: 0.870967741935484


<h2>Hasil Output</h2>

6.   Setelah melakukan tahap persiapan, pelatihan model, dan evaluasi model, maka output yang dihasilkan adalah metrik evaluasi performa model seperti akurasi, presisi, recall, dan F1-score.

Output tersebut menunjukkan bahwa model Naive Bayes yang dilatih dengan dataset Heart-Disease-Prediction berhasil mencapai akurasi sebesar 86.88%, presisi sebesar 90%, recall sebesar 84.37%, dan F1-score sebesar 87.09%. Namun, nilai-nilai tersebut dapat berbeda-beda tergantung pada pemilihan nilai k yang optimal dan dataset yang digunakan.